### ISSUE TITLE
Memory manager UnexpectedAdmissionError

### ISSUE THREAD

**Issue by champtar** [kubernetes/kubernetes#120733]
### What happened?

Dual socket server with 96 threads total (2*24*2), ~192G of RAM, cpu & memory manager static policy,  topologyManagerPolicy best-effort, 10Gi of RAM reserved on NUMA node 0, 1 core (2 threads) reserved on NUMA node 0
kubeadm config:
```
apiVersion: kubelet.config.k8s.io/v1beta1
kind: KubeletConfiguration
...
cpuManagerPolicy: static
reservedSystemCPUs: 0,48
memoryManagerPolicy: Static
reservedMemory:
  - numaNode: 0
    limits:
      # systemReserved memory + evictionHard memory.available
      memory: 10340Mi
systemReserved:
  memory: 10240Mi
topologyManagerPolicy: best-effort
```

If I try to allocate 2 Guaranteed pod with 85Gi of RAM each, one pod is admitted, and the second one fails with `UnexpectedAdmissionError` even if it would fit using memory of both NUMA nodes.
As I'm using deployments a new pod is recreated right away and you end up with tons of failed pod in `UnexpectedAdmissionError` state.

Error in the logs is:
```
E0918 12:16:32.162481 2865761 memory_manager.go:249] "Allocate error" err="[memorymanager] failed to find NUMA nodes to extend the current topology hint"
```

### What did you expect to happen?

Either the pod becomes pending, or memory from both NUMA nodes is used

### How can we reproduce it (as minimally and precisely as possible)?

Have some memory reserved on NUMA 0, launch 2 identical pods with memory limits really close to the max so 1 pod fit on NUMA 1 but the second doesn't fit on NUMA 0 (the sum fits on the server)

### Anything else we need to know?

Here the `*_manager_state`

If I launch 1 pod with  46 CPUs / 170Gi RAM, I get CPUs from 1 NUMA nodes, and memory from both NUMA nodes
<details>

```yaml
# jq . <cpu_manager_state
{
  "policyName": "static",
  "defaultCpuSet": "0,24-48,72-95",
  "entries": {
    "d6c70f6d-adbe-4901-8b82-9505171b1368": {
      "mycontainer": "1-23,49-71"
    }
  },
  "checksum": 2697499474
}

# jq . <memory_manager_state 
{
  "policyName": "Static",
  "machineState": {
    "0": {
      "numberOfAssignments": 1,
      "memoryMap": {
        "hugepages-1Gi": {
          "total": 0,
          "systemReserved": 0,
          "allocatable": 0,
          "reserved": 0,
          "free": 0
        },
        "hugepages-2Mi": {
          "total": 0,
          "systemReserved": 0,
          "allocatable": 0,
          "reserved": 0,
          "free": 0
        },
        "memory": {
          "total": 99434803200,
          "systemReserved": 10842275840,
          "allocatable": 88592527360,
          "reserved": 88592527360,
          "free": 0
        }
      },
      "cells": [
        0,
        1
      ]
    },
    "1": {
      "numberOfAssignments": 1,
      "memoryMap": {
        "hugepages-1Gi": {
          "total": 0,
          "systemReserved": 0,
          "allocatable": 0,
          "reserved": 0,
          "free": 0
        },
        "hugepages-2Mi": {
          "total": 0,
          "systemReserved": 0,
          "allocatable": 0,
          "reserved": 0,
          "free": 0
        },
        "memory": {
          "total": 101409165312,
          "systemReserved": 0,
          "allocatable": 101409165312,
          "reserved": 93943582720,
          "free": 7465582592
        }
      },
      "cells": [
        0,
        1
      ]
    }
  },
  "entries": {
    "d6c70f6d-adbe-4901-8b82-9505171b1368": {
      "mycontainer": [
        {
          "numaAffinity": [
            0,
            1
          ],
          "type": "memory",
          "size": 182536110080
        }
      ]
    }
  },
  "checksum": 382005683
}

```

</details>

If I launch a pod with 46 CPUs / 85Gi RAM, it's properly placed on NUMA 1
<details>

```yaml
# jq . <cpu_manager_state
{
  "policyName": "static",
  "defaultCpuSet": "0-23,47-71,95",
  "entries": {
    "62f6c733-dae0-4c97-a5a9-bbb5c4757a07": {
      "mycontainer": "24-46,72-94"
    }
  },
  "checksum": 3342740372
}

# jq . <memory_manager_state 
{
  "policyName": "Static",
  "machineState": {
    "0": {
      "numberOfAssignments": 0,
      "memoryMap": {
        "hugepages-1Gi": {
          "total": 0,
          "systemReserved": 0,
          "allocatable": 0,
          "reserved": 0,
          "free": 0
        },
        "hugepages-2Mi": {
          "total": 0,
          "systemReserved": 0,
          "allocatable": 0,
          "reserved": 0,
          "free": 0
        },
        "memory": {
          "total": 99434803200,
          "systemReserved": 10842275840,
          "allocatable": 88592527360,
          "reserved": 0,
          "free": 88592527360
        }
      },
      "cells": [
        0
      ]
    },
    "1": {
      "numberOfAssignments": 1,
      "memoryMap": {
        "hugepages-1Gi": {
          "total": 0,
          "systemReserved": 0,
          "allocatable": 0,
          "reserved": 0,
          "free": 0
        },
        "hugepages-2Mi": {
          "total": 0,
          "systemReserved": 0,
          "allocatable": 0,
          "reserved": 0,
          "free": 0
        },
        "memory": {
          "total": 101409165312,
          "systemReserved": 0,
          "allocatable": 101409165312,
          "reserved": 91268055040,
          "free": 10141110272
        }
      },
      "cells": [
        1
      ]
    }
  },
  "entries": {
    "62f6c733-dae0-4c97-a5a9-bbb5c4757a07": {
      "mycontainer": [
        {
          "numaAffinity": [
            1
          ],
          "type": "memory",
          "size": 91268055040
        }
      ]
    }
  },
  "checksum": 3166358723
}
```

</details>

If I try to launch 2 pods with 46 CPUs / 85Gi RAM, one fails with 
```
E0918 12:16:32.162481 2865761 memory_manager.go:249] "Allocate error" err="[memorymanager] failed to find NUMA nodes to extend the current topology hint"
```

If I try to launch 2 pods with 46 CPUs / 80Gi RAM, everything works

### Kubernetes version

1.26.8

### Cloud provider

NONE / bare-metal

### OS version

Alma 8.8 base + rpm-ostree

### Install tools

kubeadm

### Container runtime (CRI) and version (if applicable)

containerd 1.7.5

**Comment by ffromani** [comment]
/cc

Hi!

> Either the pod becomes pending, or memory from both NUMA nodes is used

Just to clarify, the memory manager can't keep the pod pending, this is a scheduler feature and builtin kube scheduler (= none of its plugin) is NUMA-aware atm. So in this case memory manager should extend the pod to both NUMA zones, and failing to do so is likely the issue

**Comment by champtar** [comment]
> Hello @champtar, Is it possible that you are hitting the exact scenario describe at the KEP: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1769-memory-manager#simulation---how-the-memory-manager-works-by-examples ? Please check the first two steps.

I was going to say I'm hitting the expected unexpected error, but I'm not sure the KEP doc 100% match the current memoryManager behavior (at least when using  `topologyManagerPolicy: best-effort`).

test 1: the "reject pod2 from the KEP" actually works: 
- start pod 60 cpus / 120Gi RAM
- wait
- start pod 20 cpus / 20Gi RAM

<details>

```
# jq . <memory_manager_state
...
  "entries": {
    "c32d8509-dcb4-4eb9-b5b2-a08f98c5174d": {
      "mycontainer": [
        {
          "numaAffinity": [
            0,
            1
          ],
          "type": "memory",
          "size": 128849018880
        }
      ]
    },
    "fc5b236c-041f-4dfc-b615-7dc6cb5456c4": {
      "mycontainer": [
        {
          "numaAffinity": [
            0,
            1
          ],
          "type": "memory",
          "size": 21474836480
        }
      ]
    }
  },
```

</details>

test 2 (start a 1 NUMA node container then one than spans 2 NUMA node) also works:
- start pod 20 cpus / 20Gi RAM
- wait
- start pod 60 cpus / 120Gi RAM

<details>

```
# jq . <memory_manager_state
...
    "785f9d33-7cc3-43f0-8863-66cd80e8c7ff": {
      "mycontainer": [
        {
          "numaAffinity": [
            0
          ],
          "type": "memory",
          "size": 21474836480
        }
      ]
    },
    "c37b0288-cc74-49cd-82fb-d5f5afa3f98e": {
      "mycontainer": [
        {
          "numaAffinity": [
            0,
            1
          ],
          "type": "memory",
          "size": 128849018880
        }
      ]
    }
  },
```

</details>

What doesn't work is when you get too close to the limits it seems

**Comment by Tal-or** [comment]
> > Hello @champtar, Is it possible that you are hitting the exact scenario describe at the KEP: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1769-memory-manager#simulation---how-the-memory-manager-works-by-examples ? Please check the first two steps.
> 
> I was going to say I'm hitting the expected unexpected error, but I'm not sure the KEP doc 100% match the current memoryManager behavior (at least when using `topologyManagerPolicy: best-effort`).
> 
> test 1: the "reject pod2 from the KEP" actually works:
> 
> * start pod 60 cpus / 120Gi RAM
> * wait
> * start pod 20 cpus / 20Gi RAM
> 
> test 2 (start a 1 NUMA node container then one than spans 2 NUMA node) also works:
> 
> * start pod 20 cpus / 20Gi RAM
> * wait
> * start pod 60 cpus / 120Gi RAM
> 
> What doesn't work is when you get too close to the limits it seems

Interesting 🤔 those pods are guaranteed I presume?

And what about the total node memory capacity?
When you check the node status do you see you have enough memory left to fulfill the second pod request?

**Comment by champtar** [comment]
@Tal-or click on details you will have the content of `*_manager_state`
yes they are guaranteed else they would not show up in memory_manager_state
it's less then the total node capacity else pod would be pending

**Comment by ffromani** [comment]
assigned to me for bookkeeping reason but @Tal-or is doing the real work here

**Comment by Tal-or** [comment]
Hello @champtar, so far I didn't manage to reproduce the issue you're describing.
In my case I manage to ran both pods, or end up with the second one on `pending`.

However, I managed to find something similar (different affinity error) but with `none` topology manager policy and not `best-effort` as you described which I'm going to open a separate issue for.

Could you please share the journal logs while you're reproducing the issue again?
Basically on the node where the pod is about to land, share the `journalctl -u kubelet` output

**Comment by Tal-or** [comment]
One more thing please share `kubectl describe node` output of the node before and after you reproduce the issue

**Comment by champtar** [comment]
@Tal-or off until Tuesday, I'll try to reproduce with 1.28.2 then and share the logs next week

**Comment by Tal-or** [comment]
NP. I won't be here next week either, so no rush

**Comment by Tal-or** [comment]
Hello @champtar did you manage to reproduce the issue?

**Comment by champtar** [comment]
@Tal-or yes, reproduced on my dev machine in 1.28.2, tell me what else you need / if you want me to rebuild kubelet with some debug that's also possible
[01-before.txt](https://github.com/kubernetes/kubernetes/files/12911103/01-before.txt)
[02-1pod.txt](https://github.com/kubernetes/kubernetes/files/12911104/02-1pod.txt)
[03-2pods.txt](https://github.com/kubernetes/kubernetes/files/12911105/03-2pods.txt)
[kubelet_logs.txt](https://github.com/kubernetes/kubernetes/files/12911106/kubelet_logs.txt)

**Comment by Tal-or** [comment]
Hi @champtar,

I did mange to reproduce the issue, but following the code it seems exactly like I mentioned in my previous [comments](https://github.com/kubernetes/kubernetes/issues/120733#issuecomment-1725078070)

I'll try to elaborate:
When the first pod gets admitted it placed on NUMA 0 (in my case). 
When the second pod gets admitted, there is not enough memory on each NUMA individually (NUMA 0 **or** NUMA1 to satisfy the memory request,  so memory manager tries to span the memory allocation through NUMA 0 **and** NUMA 1).

But it failed short in this line:
https://github.com/kubernetes/kubernetes/blob/01b96e7704a32efd5627769fce1d5bbcd9f4472d/pkg/kubelet/cm/memorymanager/policy_static.go#L492 because NUMA node 0 is now used for single-NUMA memory allocation
which align with the KEP .

In such case, the pod cannot move to `pending` because this is done at the scheduler level, while from the scheduler's POV there is enough memory on the node to satisfy the request. 

Now, about your comment here: https://github.com/kubernetes/kubernetes/issues/120733#issuecomment-1726333248

> test 1: the "reject pod2 from the KEP" actually works:
>
>start pod 60 cpus / 120Gi RAM
>wait
>start pod 20 cpus / 20Gi RAM

Make sense because both NUMA 0 and NUMA 1 are already serves mulit-NUMA allocation, since the first pod spans across the two NUMAs, the second pod can squeezed into one of them with not problem.

> test 2 (start a 1 NUMA node container then one than spans 2 NUMA node) also works:
>
>start pod 20 cpus / 20Gi RAM
>wait
>start pod 60 cpus / 120Gi RAM

This case is more interesting and I didn't manage to reproduce it - I always got `UnexpectedAdmissionError` which is align with the KEP.

Could you please increase Kubelet logs verbosity, reproduce test 2 and attached the following information:
1. Kubelet logs
2. memory_state file before
3. pod1 manifest
4. memory_state file after pod1
5. pod2 manifest
6. memory_state file after pod2

Thanks

**Comment by champtar** [comment]
Hi @Tal-or 
this is a different server with a bit less CPUs, but test 2 still works for me
for verbosity I've used v=99 instead of v=2

Here part of my KubeletConfiguration
```
cgroupDriver: systemd
cpuManagerPolicy: static
reservedSystemCPUs: 0,36
memoryManagerPolicy: Static
reservedMemory:
  - numaNode: 0
    limits:
      # systemReserved memory + evictionHard memory.available
      memory: 10340Mi
systemReserved:
  memory: 10240Mi
topologyManagerPolicy: best-effort
```

The logs:
[01-before.txt](https://github.com/kubernetes/kubernetes/files/13302660/01-before.txt)
[02-1pod.txt](https://github.com/kubernetes/kubernetes/files/13302661/02-1pod.txt)
[03-2pods.txt](https://github.com/kubernetes/kubernetes/files/13302662/03-2pods.txt)
[kubelet_logs.txt](https://github.com/kubernetes/kubernetes/files/13302663/kubelet_logs.txt)

And the commands:
<details>

```
k8s_state > 01-before.txt

kubectl apply -f- <<'EOF'
apiVersion: apps/v1
kind: Deployment
metadata:
  name: test1
spec:
  replicas: 1
  strategy:
    type: Recreate
  selector:
    matchLabels:
      test: '1'
  template:
    metadata:
      labels:
        test: '1'
    spec:
      containers:
      - name: test1
        image: 127.0.0.1:5000/pause:3.9
        resources:
          limits:
            cpu: 20
            memory: 20Gi
EOF

k8s_state > 02-1pod.txt

kubectl apply -f- <<'EOF'
apiVersion: apps/v1
kind: Deployment
metadata:
  name: test2
spec:
  replicas: 1
  strategy:
    type: Recreate
  selector:
    matchLabels:
      test: '2'
  template:
    metadata:
      labels:
        test: '2'
    spec:
      containers:
      - name: test2
        image: 127.0.0.1:5000/pause:3.9
        resources:
          limits:
            cpu: 40
            memory: 120Gi
EOF

k8s_state > 03-2pods.txt

journalctl -b0 --unit kubelet.service > kubelet_logs.txt
```
</details>

**Comment by champtar** [comment]
@Tal-or I think I pinpointed a bit more the issue
```
policy_static.go:522] "TopologyHints generated" pod="default/test2-7d8d5f7b6-2q8df" containerName="test2" cpuHints=[{"NUMANodeAffinity":3,"Preferred":true}]
```
For my test2 pod the cpu manager generated a hint with both NUMA nodes.

For test2 pod, if CPU request span both NUMA nodes, then the memory allocation will work whether it requires to use 1 or 2 NUMA nodes
If CPU request only requires 1 NUMA node, then it'll fail if memory allocation need to extend to 2 NUMA node.

**Comment by Tal-or** [comment]
Hello @champtar,

After carefully going through the code and the KEP I think that the behavior you're expecting:
_"Either the pod becomes pending, or memory from both NUMA nodes is used"_ is not aligned with the architecture.

Let's break it down:
_"The pod becomes pending"_ - From the default scheduler POV, the node have enough resources (because the scheduler is not NUMA aware) so the fact that the pod scheduled on this node is align with the expected behavior.
**Sidenote:** if you want to enable NUMA aware scheduling you can take a look at https://github.com/k8stopologyawareschedwg/documentation - this is a working group which I'm part of, and we  developed a mechanism that supports NUMA aware scheduling.

_"memory from both NUMA nodes is used"_ - Because NUMA 0 is already serves as single NUMA memory allocation, it cannot serves as multi NUMA memory allocation as well, otherwise the memory pinning is not guaranteed since linux kernel doesn't support multi-NUMA allocation (pinning).

And you were right, when the hint generated by cpu manager, spans for both NUMAs the pod gets admitted, and that's exactly where **actual** bug is -  the pod should never get admitted.
IOW you should always get the error: 
```
E0918 12:16:32.162481 2865761 memory_manager.go:249] "Allocate error" err="[memorymanager] failed to find NUMA nodes to extend the current topology hint"
```
So the fix here should be preventing the pod admission even when the CPU manager hint spans across both NUMAs.

Now, in your case, if you still want and need both pods to get admitted you can change the `memoryManagerPolicy` from `Static` to `None` because as I mentioned, the memory pinning in your scenario is not guaranteed anyway.

**Comment by Tal-or** [comment]
Hello @champtar ,

After carefully going through the code and the KEP I think that the behavior you're expecting:
_"Either the pod becomes pending, or memory from both NUMA nodes is used"_ is not aligned with the architecture.

Let's break it down:
_"The pod becomes pending"_ - From the default scheduler POV, the node have enough resources (because the scheduler is not NUMA aware) so the fact that the pod scheduled on this node is align with the expected behavior.
**Sidenote:** if you want to enable NUMA aware scheduling you can take a look at https://github.com/k8stopologyawareschedwg/documentation

_"memory from both NUMA nodes is used"_ - Because NUMA 0 is already serves as single NUMA memory allocation, it cannot serves as multi NUMA memory allocation as well, otherwise the memory pinning is not guaranteed since linux kernel doesn't support multi-NUMA allocation (pinning). What it can do is grouping NUMA 0 with NUMA 1 into a single group but again this is against the original design: 
Please check https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1769-memory-manager#simulation---how-the-memory-manager-works-by-examples and look for "No spare capacity, reject Pod 7" (last example)

And you were right, when the hint generated by cpu manager, spans for both NUMAs the pod gets admitted, and that's exactly where actual bug is - the pod should never get admitted.
IOW you should always get the error:
```
E0918 12:16:32.162481 2865761 memory_manager.go:249] "Allocate error" err="[memorymanager] failed to find NUMA nodes to extend the current topology hint"
```

So the fix here should be preventing the pod admission even when the CPU manager hint spans across both NUMAs.
Now, in your case, if you still want and need both pods to get admitted you can change the `memoryManagerPolicy` from `Static` to `None` because as I mentioned, the memory pinning in your scenario is not guaranteed anyway.

**Comment by champtar** [comment]
Hello @Tal-or,

I still think the bug is that memoryManager doesn't respect topologyManagerPolicy: best-effort.
For now I'm exploring using a simple NRI plugin to set the cpuset.mems to the nodes of cpuset.cpus.

**Comment by Tal-or** [comment]
@champtar ,
The best-effort means that even if the affinity is not preferred, the pod would get admit anyway.

But here we're talking about respecting the NUMA node grouping done by memory manager.
This is an architectural design we should refer to the original creator memory manager or at least to topology maintainers folks.

@ffromani do you know who can answer on such questions?

**Comment by Tal-or** [comment]
@champtar some additional info about the limitation i mentioned above https://kubernetes.io/blog/2021/08/11/kubernetes-1-22-feature-memory-manager-moves-to-beta/#single-vs-cross-numa-node-allocation
